### DISCUSSION TITLE
Encountered a bug when calculating the steady state

### DISCUSSION THREAD

**Discussion by skothegh** [netket/netket#820] [General]
Hola,

when calculating the steady state for a simple toy model using the Adam optimizer I found that while LdagL converged towards 0 fairly quickly, I tended to get density matrices that differed wildly and randomly from the expected density matrices.

To get a better clue of what was going on, I loaded the resulting density matrix into a QUTIP script and calculated LdagL exactly and found that it was usually not zero.

When calculating all expectation values at every iteration (ss.run(Niter=1) for 50 steps) I found that it always perfectly converged against the correct density matrix. When I only measured at the end of the calculation, the density matrix was wrong again.

Below you find the minimum example. The Lindbladian only contains a dissipation part where all jump operators are sigmam(). The Hamiltonian is 0. This should result in a simple steady state with all spins pointing down.

The script calculates the steady state density matrix three times, onnce "in one go" (Niter=50), once without measuring anything and once measuring sigmax(0) at each step. At the end of each calculation it measures sigmaz(0). You will find that the last calculation always correctly estimates sz=-1, while the first two calculations  can result in anything (including -1 sometimes).

```
import netket as nk
import netket.legacy as nkl
import numpy as np

# System Lattic
L = 4
g = nk.graph.Hypercube(length=L, n_dim = 1, pbc = False)

# Hilbert space on the Lattice
hi = nk.hilbert.Spin(s= 1/2, N=g.n_nodes)

# Hamiltonian and parameters
J = 0
h = 0

km = 1
kp = 0

ha = nk.operator.LocalOperator(hi)
for i in range(L):
    ha += h * nk.operator.spin.sigmaz(hi,i)

for i in range(L-1):
    ha += (
          J
          * nk.operator.spin.sigmax(hi,i)
          * nk.operator.spin.sigmax(hi,i+1)
          )

# Collapse operators and parameters
c_ops = []
for i in range(L):
    c_ops.append( km * nk.operator.spin.sigmam(hi, i) )
    c_ops.append( kp * nk.operator.spin.sigmap(hi, i) )

# Lindblad operator
lind = nk.operator.LocalLiouvillian(ha, c_ops)

# model function
ma = nk.models.NDM(beta=1, alpha=1)

# sampler
sa = nk.sampler.MetropolisLocal(lind.hilbert)

# optimizer
op = nk.optimizer.Adam(learning_rate=0.05)
sr = None#nk.optimizer.SR(diag_shift=0.01)

# variational state
vs = nk.variational.MCMixedState(sa, ma, n_samples=5000, n_samples_diag=5000)


# "in one go"
vs.init_parameters(nk.nn.initializers.normal(stddev=0.001))
ss = nk.SteadyState(lind, op, variational_state = vs, preconditioner = sr)
ss.run(n_iter=50, out="test")
print(ss.estimate(nk.operator.spin.sigmaz(hi,0)))

# without measuring each step
vs.init_parameters(nk.nn.initializers.normal(stddev=0.001))
ss = nk.SteadyState(lind, op, variational_state = vs, preconditioner = sr)

for i in range(50):
    ss.run(n_iter=1, out="test")
print(ss.estimate(nk.operator.spin.sigmaz(hi,0)))


# With measuring something each step
vs.init_parameters(nk.nn.initializers.normal(stddev=0.001))
ss = nk.SteadyState(lind, op, variational_state = vs, preconditioner = sr)
for i in range(50):
    ss.run(n_iter=1, out="test")
    ss.estimate(nk.operator.spin.sigmax(hi, 0))

print(ss.estimate(nk.operator.spin.sigmaz(hi,0)))

```

**Comment by PhilipVinc** [comment]
> using the Adam optimizer

Adam + SR is a recipe for disaster. Use at your own risk. SR should be used with SGD. Maybe momentum if you are very careful.

> To get a better clue of what was going on, I loaded the resulting density matrix into a QUTIP script and calculated LdagL exactly and found that it was usually not zero.

QuTiP matrices stay to NetKet matrices as UK Cars stay to EU cars. It would be lovely if UK adapted to the standard of the rest of the world instead of exporting their weird standard to their colonies.
In particular, QuTiP encodes density matrices with Column Major (this comes from their matlab heirtage) while NetKet adapts to the python/numpy standard of row major. so there's always a transposition in between. 
I think we also have different ordering of the kronecker product so you shouldn't really expect to take netket's steady-state and plug it into qutip's liouvillian to copmute l dag l.

I don't understand why you wouldn't do this in netket

```python
>>> import jax.numpy as jnp
>>> jnp.sum(jnp.abs(lind.to_sparse()@vs.to_matrix().reshape(-1))**2)
DeviceArray(1.43480886e-05, dtype=float64)
```

which is correct (this is after running your script).
Alternatively, you can also compute yourself the expectation value of LdagL, which gives you 
```python
>>> vs.expect(lind.H@lind)
1.085e-09 ± 3.329e-27 [σ²=1.193e-49, R̂=0.9984]
```
that is underestimated because the sampling is wrong. If we increase the samples you can get the good result:
```python
>>> vs.n_samples=50000
>>> vs.expect(lind.H@lind)
0.0000050 ± 0.0000050 [σ²=0.0000012, R̂=1.0000]
```

> When calculating all expectation values at every iteration (ss.run(Niter=1) for 50 steps) I found that it always perfectly converged against the correct density matrix. When I only measured at the end of the calculation, the density matrix was wrong again.

I find this very weird because those three things all share the same code path. 
I just ran your code and all three blocks give me the same result. 
I also tried to run your code by setting the seed before every _run_, both the one used to inialize the parameters and the master one used to initialize the samplers.
I get exactly the same parameters in all three runs so i don't see whatyou are describing.

```python
# variational state
vs = nk.vqs.MCMixedState(sa, ma, n_samples=5000, n_samples_diag=5000, seed=123)
vs.init_parameters(nk.nn.initializers.normal(stddev=0.001), seed=123)

ss = nk.SteadyState(lind, op, variational_state = vs, preconditioner = sr)
log1 = nk.logging.RuntimeLog()
ss.run(n_iter=50, out=log1, obs={'z':nk.operator.spin.sigmaz(hi,0)})
```

```python
# variational state
vs2 = nk.vqs.MCMixedState(sa, ma, n_samples=5000, n_samples_diag=5000, seed=123)
vs2.init_parameters(nk.nn.initializers.normal(stddev=0.001), seed=123)

ss = nk.SteadyState(lind, op, variational_state = vs2, preconditioner = sr)
data = []
for i in tqdm(range(50)):
    ss.run(n_iter=1, out="test", show_progress=False)
    data.append(ss.estimate(nk.operator.spin.sigmaz(hi, 0)).Mean)

print(vs.expect(nk.operator.spin.sigmaz(hi,0)))
```
and I check that parameters are the same by doing
```python
import jax
import numpy as np
print(jax.tree_multimap(lambda a,b: jnp.allclose(a,b), vs.parameters, vs2.parameters))
```

I'm running the latest release of netket (3.0b3), but this stuff has been stable for a while now..

**Reply by skothegh** [reply]
Hi, I am not using SR. Its commented out because I regularly switch between Sgd and Adam. I forgot to delete the comment for the minimum example. 

The transposition is taken care off. If the code returns the correct density matrix I also get the correct result in QUTIP. I tend to use QUTIP for exact calculations because I am more firm with that and I didn't know about the Netket options.

You are right, it was undersampled. Thank you for your explanation and I am sorry for taking your time.

**Comment by VolodyaCO** [comment]
Hi, haven't checked this in detail, but what happens is that the ML model
from netket does not evaluate all the matrix elements of the density matrix
(or the ground state vector in the Hamiltonian case). Instead, it restricts
itself to a corner of the Hilbert space. Thus, it uses those matrix
elements to compute expected values. However, it can be the case that
matrix elements that do not come in the sample are wrongly estimated by the
neural quantum state.

I witnessed this while working on the Bose-Hubbard Hamiltonian in the
following paper: https://journals.jps.jp/doi/abs/10.7566/JPSJ.89.094002

This is essentially a consequence of learning the wave function in the
corner of the Hilbert space that is relevant for the ground or the
steady-state. However, outside this Hilbert space corner, you are not
guaranteed to have a correct learning of the wavefunction.


On Wed, 14 Jul 2021 at 09:34, skothegh ***@***.***> wrote:

> Hi, I am not using SR. Its commented out because I regularly switch
> between Sgd and Adam. I forgot to delete the comment for the minimum
> example.
>
> The transposition is taken care off. If the code returns the correct
> density matrix I also get the correct result in QUTIP. I tend to use QUTIP
> for exact calculations because I am more firm with that and I didn't know
> about the Netket options.
>
> That you get the same results for all three is incredibly weird. I only
> reported what I saw.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/netket/netket/discussions/820#discussioncomment-1004238>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHQJA7Z76TIZSNRDCZYUSK3TXWOADANCNFSM5ALPJS7Q>
> .
>


-- 
Best wishes,

Vladimir Vargas-Calderón
<https://www.researchgate.net/profile/Vladimir_Vargas-Calderon>
PhD Physics Student @ Universidad Nacional de Colombia

-- 
*Aviso legal:* El contenido de este mensaje y los archivos adjuntos son 
confidenciales y de uso exclusivo de la Universidad Nacional de Colombia. 
Se encuentran dirigidos sólo para el uso del destinatario al cual van 
enviados. La reproducción, lectura y/o copia se encuentran prohibidas a 
cualquier persona diferente a este y puede ser ilegal. Si usted lo ha 
recibido por error, infórmenos y elimínelo de su correo. Los Datos 
Personales serán tratados conforme a la Ley 1581 de 2012 y a nuestra 
Política de Datos Personales que podrá consultar en la página web 
www.unal.edu.co <http://www.unal.edu.co/>.* *Las opiniones, informaciones, 
conclusiones y cualquier otro tipo de dato contenido en este correo 
electrónico, no relacionados con la actividad de la Universidad Nacional de 
Colombia, se entenderá como personales y de ninguna manera son avaladas por 
la Universidad.
